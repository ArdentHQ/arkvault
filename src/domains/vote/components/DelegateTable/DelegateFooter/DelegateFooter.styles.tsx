import tw, { styled } from "twin.macro";

import { Circle } from "@/app/components/Circle";

const disabledColor = tw`text-theme-secondary-500 dark:text-theme-secondary-700`;

export const LabelWrapper = styled.div`
	${tw`text-sm leading-tighter`};
	${disabledColor};
`;

export const TextWrapper = styled.div<{ disabled?: boolean }>`
	${tw`text-lg leading-tighter`};

	${({ disabled }) => {
		if (disabled) {
			return disabledColor;
		}

		return tw`text-theme-text`;
	}}
`;

export const StyledCircle = styled(Circle)<{ disabled?: boolean }>`
	${tw`bg-theme-background`};

	${({ disabled }) => {
		if (disabled) {
			return [tw`border-theme-secondary-500 dark:border-theme-secondary-700`, disabledColor];
		}

		return tw`border-theme-secondary-900 text-theme-secondary-900 dark:(border-theme-secondary-600 text-theme-secondary-600)`;
	}}
`;
