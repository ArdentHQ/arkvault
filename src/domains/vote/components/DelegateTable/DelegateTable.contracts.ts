import { Networks } from "@ardenthq/sdk";
import { Contracts } from "@ardenthq/sdk-profiles";
import React from "react";

export interface DelegateTableProperties {
	delegates: Contracts.IReadOnlyWallet[];
	isLoading?: boolean;
	maxVotes: number;
	unvoteDelegates: VoteDelegateProperties[];
	voteDelegates: VoteDelegateProperties[];
	selectedWallet: Contracts.IReadWriteWallet;
	votes: Contracts.VoteRegistryItem[];
	resignedDelegateVotes?: Contracts.VoteRegistryItem[];
	onContinue?: (unvotes: VoteDelegateProperties[], votes: VoteDelegateProperties[]) => void;
	isCompact?: boolean;
	subtitle?: React.ReactNode;
	searchQuery?: string;
	activeWallet?: Contracts.IReadWriteWallet;
}

export interface VoteDelegateProperties {
	delegateAddress: string;
	amount: number;
}

export interface DelegateTableColumnsProperties {
	network: Networks.Network;
	isLoading?: boolean;
}
