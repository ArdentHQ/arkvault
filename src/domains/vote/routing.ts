import { ProfilePaths } from "@/router/paths";
import { RouteItem } from "@/router/router.types";
import preloadLazy from "@/utils/preload-lazy";

const Votes = preloadLazy(() => import("./pages/Votes"));

export const VoteRoutes: RouteItem[] = [
	{
		component: Votes,
		exact: true,
		path: ProfilePaths.Votes,
	},
	{
		component: Votes,
		exact: true,
		path: ProfilePaths.VotesWallet,
	},
];
