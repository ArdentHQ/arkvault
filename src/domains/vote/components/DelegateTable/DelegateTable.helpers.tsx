import { Contracts } from "@ardenthq/sdk-profiles";
import React, { useMemo } from "react";
import { useTranslation } from "react-i18next";
import { Column } from "react-table";

import { DelegateTableColumnsProperties, VoteDelegateProperties } from "./DelegateTable.contracts";
import { Icon } from "@/app/components/Icon";
import { Tooltip } from "@/app/components/Tooltip";

export const useDelegateTableColumns = ({ network, isLoading }: DelegateTableColumnsProperties) => {
	const { t } = useTranslation();

	return useMemo<Column<Contracts.IReadOnlyWallet>[]>(() => {
		const templateColumns: Column<Contracts.IReadOnlyWallet>[] = [
			{
				Header: (
					<>
						<span className="hidden sm:inline">{t("COMMON.RANK")}</span>
						<span className="sm:hidden">#</span>
					</>
				),
				accessor: "rank",
				minimumWidth: true,
			},
			{
				Header: t("VOTE.DELEGATE_TABLE.NAME"),
				accessor: (delegate) => isLoading || delegate.username(),
				className: "justify-start",
				headerClassName: "w-3/4 sm:w-auto",
			},
			{
				Header: t("COMMON.STATUS"),
				accessor: "status",
				className: "justify-center",
				disableSortBy: true,
				headerClassName: "hidden sm:table-cell",
				minimumWidth: true,
			},
			{
				Header: t("COMMON.EXPLORER"),
				accessor: (delegate) => isLoading || delegate.explorerLink(),
				className: "justify-center",
				disableSortBy: true,
				headerClassName: "hidden sm:table-cell",
				minimumWidth: true,
			},
		];

		if (network.votesAmountMinimum() > 0) {
			templateColumns.push({
				Header: (
					<div className="flex items-center space-x-3 px-3">
						<p>{t("VOTE.DELEGATE_TABLE.VOTE_AMOUNT.TITLE")}</p>
						<Tooltip
							content={t("VOTE.DELEGATE_TABLE.VOTE_AMOUNT.TOOLTIP", {
								coinId: network.coin(),
							})}
						>
							<span className="rounded-full bg-theme-primary-100 p-1 text-theme-primary-600 dark:bg-theme-secondary-800 dark:text-theme-secondary-200">
								<Icon name="QuestionMarkSmall" size="sm" />
							</span>
						</Tooltip>
					</div>
				),
				accessor: () => "voteAmount",
				className: "justify-end",
				disableSortBy: true,
				id: "voteAmount",
			} as Column<Contracts.IReadOnlyWallet>);
		}

		templateColumns.push({
			Header: t("VOTE.DELEGATE_TABLE.VOTE"),
			accessor: () => "onSelect",
			className: "justify-end",
			disableSortBy: true,
			id: "onSelect",
		});

		return templateColumns;
	}, [t, network, isLoading]);
};

export const delegateExistsInVotes = (
	votes: VoteDelegateProperties[],
	address: string,
): VoteDelegateProperties | undefined => votes.find(({ delegateAddress }) => delegateAddress === address);
