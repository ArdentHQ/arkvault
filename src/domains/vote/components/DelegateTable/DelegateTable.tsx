import { Contracts } from "@payvo/sdk-profiles";
import React, { FC, useCallback, useEffect, useMemo, useState } from "react";
import { useTranslation } from "react-i18next";

import { DelegateFooter } from "./DelegateFooter";
import { DelegateRow } from "./DelegateRow";
import { DelegateTableProperties, VoteDelegateProperties } from "./DelegateTable.contracts";
import { delegateExistsInVotes, useDelegateTableColumns } from "./DelegateTable.helpers";
import { Table } from "@/app/components/Table";
import { Pagination } from "@/app/components/Pagination";
import { EmptyResults } from "@/app/components/EmptyResults";
import { useBreakpoint } from "@/app/hooks";

const DELEGATES_PER_PAGE = 50;

export const DelegateTable: FC<DelegateTableProperties> = ({
	delegates,
	isLoading = false,
	maxVotes,
	unvoteDelegates,
	voteDelegates,
	selectedWallet,
	votes,
	resignedDelegateVotes,
	onContinue,
	isCompact: isCompactOption,
	subtitle,
	searchQuery,
}) => {
	const { t } = useTranslation();
	const [currentPage, setCurrentPage] = useState(1);
	const [selectedUnvotes, setSelectedUnvotes] = useState<VoteDelegateProperties[]>(unvoteDelegates);
	const [selectedVotes, setSelectedVotes] = useState<VoteDelegateProperties[]>(voteDelegates);
	const [isVoteDisabled, setIsVoteDisabled] = useState(false);
	const [availableBalance, setAvailableBalance] = useState(selectedWallet.balance());
	const { isXs, isSm, isMd } = useBreakpoint();
	const isCompact = useMemo(() => isCompactOption || isXs || isSm || isMd, [isCompactOption, isXs, isSm, isMd]);

	const columns = useDelegateTableColumns({ isLoading, network: selectedWallet.network() });

	const totalDelegates = useMemo(() => delegates.length, [delegates.length]);
	const hasMoreDelegates = useMemo(() => totalDelegates > DELEGATES_PER_PAGE, [totalDelegates]);
	const hasVotes = votes.length > 0;

	useEffect(() => {
		if (voteDelegates.length === 0) {
			return;
		}

		let totalVotesAmount = 0;

		for (const delegate of voteDelegates) {
			totalVotesAmount += delegate.amount;
		}

		setAvailableBalance(availableBalance - totalVotesAmount);
	}, []); // eslint-disable-line react-hooks/exhaustive-deps

	useEffect(() => {
		if ((hasVotes || maxVotes > 1) && selectedVotes.length === maxVotes) {
			setIsVoteDisabled(true);
		} else {
			setIsVoteDisabled(false);
		}
	}, [hasVotes, maxVotes, selectedVotes]);

	useEffect(() => window.scrollTo({ behavior: "smooth", top: 0 }), [currentPage]);

	useEffect(() => {
		if (!resignedDelegateVotes?.length) {
			return;
		}

		for (const { wallet } of resignedDelegateVotes) {
			if (delegateExistsInVotes(selectedUnvotes, wallet!.address())) {
				continue;
			}

			toggleUnvotesSelected(wallet!.address());
		}
	}, [resignedDelegateVotes, selectedUnvotes]); // eslint-disable-line react-hooks/exhaustive-deps

	useEffect(() => {
		setCurrentPage(1);
	}, [searchQuery]); // eslint-disable-line react-hooks/exhaustive-deps

	const toggleUnvotesSelected = useCallback(
		(address: string, voteAmount?: number) => {
			let unvotesInstance = selectedUnvotes;
			const delegateAlreadyExists = delegateExistsInVotes(selectedUnvotes, address);

			if (delegateAlreadyExists) {
				unvotesInstance = selectedUnvotes.filter(({ delegateAddress }) => delegateAddress !== address);
			}

			if (delegateAlreadyExists && voteAmount === undefined) {
				setSelectedUnvotes(unvotesInstance);

				if (maxVotes === 1 && selectedVotes.length > 0) {
					setSelectedVotes([]);
				}

				return;
			}

			const voteDelegate: VoteDelegateProperties = {
				amount: voteAmount ?? 0,
				delegateAddress: address,
			};

			const delegate = votes.find(({ wallet }) => wallet?.address() === address);
			if (delegate?.amount && voteAmount === undefined) {
				voteDelegate.amount = delegate.amount;
			}

			if (maxVotes === 1) {
				setSelectedUnvotes([voteDelegate]);
			} else {
				setSelectedUnvotes([...unvotesInstance, voteDelegate]);
			}
		},
		[selectedUnvotes, votes, setSelectedUnvotes, setSelectedVotes, maxVotes, selectedVotes.length],
	);

	const toggleVotesSelected = useCallback(
		(address: string, voteAmount?: number) => {
			let votesInstance = selectedVotes;
			const delegateAlreadyExists = delegateExistsInVotes(selectedVotes, address);

			if (delegateAlreadyExists) {
				votesInstance = selectedVotes.filter(({ delegateAddress }) => delegateAddress !== address);
			}

			if (delegateAlreadyExists && voteAmount === undefined) {
				setSelectedVotes(votesInstance);

				if (maxVotes === 1 && hasVotes) {
					setSelectedUnvotes([]);
				}

				return;
			}

			const voteDelegate: VoteDelegateProperties = {
				amount: voteAmount ?? 0,
				delegateAddress: address,
			};

			if (maxVotes === 1) {
				setSelectedVotes([voteDelegate]);

				if (hasVotes) {
					setSelectedUnvotes(
						votes.map((vote) => ({
							amount: vote.amount,
							delegateAddress: vote.wallet!.address(),
						})),
					);
				}
			} else {
				setSelectedVotes([...votesInstance, voteDelegate]);
			}
		},
		[selectedVotes, hasVotes, setSelectedUnvotes, maxVotes, votes],
	);

	const handleSelectPage = useCallback(
		(page: number) => {
			setCurrentPage(page);
		},
		[setCurrentPage],
	);

	const showSkeleton = useMemo(() => totalDelegates === 0 && isLoading, [totalDelegates, isLoading]);
	const tableData = useMemo<Contracts.IReadOnlyWallet[]>(() => {
		if (!showSkeleton) {
			return delegates;
		}

		return Array.from<Contracts.IReadOnlyWallet>({ length: DELEGATES_PER_PAGE }).fill(
			{} as Contracts.IReadOnlyWallet,
		);
	}, [delegates, showSkeleton]);

	const renderTableRow = useCallback(
		(delegate: Contracts.IReadOnlyWallet, index: number) => {
			let voted: Contracts.VoteRegistryItem | undefined;

			if (hasVotes) {
				// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
				voted = votes.find(({ wallet }) => wallet?.address() === delegate?.address?.());
			}

			return (
				<DelegateRow
					index={index}
					delegate={delegate}
					selectedUnvotes={selectedUnvotes}
					selectedVotes={selectedVotes}
					selectedWallet={selectedWallet}
					availableBalance={availableBalance}
					setAvailableBalance={setAvailableBalance}
					voted={voted}
					isVoteDisabled={isVoteDisabled}
					isLoading={showSkeleton}
					isCompact={isCompact}
					toggleUnvotesSelected={toggleUnvotesSelected}
					toggleVotesSelected={toggleVotesSelected}
				/>
			);
		},
		[
			votes,
			selectedUnvotes,
			selectedVotes,
			selectedWallet,
			availableBalance,
			setAvailableBalance,
			isVoteDisabled,
			showSkeleton,
			isCompact,
			toggleUnvotesSelected,
			toggleVotesSelected,
			hasVotes,
		],
	);

	if (!isLoading && totalDelegates === 0) {
		return (
			<EmptyResults
				className="mt-16"
				title={t("COMMON.EMPTY_RESULTS.TITLE")}
				subtitle={t("VOTE.VOTES_PAGE.NO_RESULTS")}
			/>
		);
	}

	return (
		<div data-testid="DelegateTable">
			<h2 className="mb-6 hidden text-lg font-bold md:block">{t("VOTE.DELEGATE_TABLE.TITLE")}</h2>

			{!!subtitle && subtitle}

			<Table columns={columns} data={tableData} rowsPerPage={DELEGATES_PER_PAGE} currentPage={currentPage}>
				{renderTableRow}
			</Table>

			<div className="mt-8 flex w-full justify-center">
				{hasMoreDelegates && (
					<Pagination
						totalCount={totalDelegates}
						itemsPerPage={DELEGATES_PER_PAGE}
						currentPage={currentPage}
						onSelectPage={handleSelectPage}
					/>
				)}
			</div>

			<DelegateFooter
				selectedWallet={selectedWallet}
				availableBalance={availableBalance}
				selectedVotes={selectedVotes}
				selectedUnvotes={selectedUnvotes}
				maxVotes={maxVotes}
				onContinue={onContinue}
			/>
		</div>
	);
};
