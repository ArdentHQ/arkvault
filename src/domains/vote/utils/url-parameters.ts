import { VoteDelegateProperties } from "@/domains/vote/components/DelegateTable/DelegateTable.contracts";

export type ParameterNameProperties = "vote" | "unvote";

export const appendParameters = (
	parameters: URLSearchParams,
	parameterName: ParameterNameProperties,
	votes: VoteDelegateProperties[],
) => {
	for (const { delegateAddress, amount } of votes) {
		parameters.append(parameterName, `${delegateAddress}, ${amount}`);
	}
};

export const getParameters = (
	parameters: URLSearchParams,
	parameterName: ParameterNameProperties,
): VoteDelegateProperties[] =>
	parameters.getAll(parameterName)?.map((vote) => {
		const voteDelegate = vote.split(",");

		return {
			amount: +voteDelegate[1],
			delegateAddress: voteDelegate[0],
		};
	});
