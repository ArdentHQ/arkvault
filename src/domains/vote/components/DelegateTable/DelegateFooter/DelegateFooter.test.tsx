import { Contracts, ReadOnlyWallet } from "@payvo/sdk-profiles";
import userEvent from "@testing-library/user-event";
import React from "react";

import { DelegateFooter } from "./DelegateFooter";
import { buildTranslations } from "@/app/i18n/helpers";
import { VoteDelegateProperties } from "@/domains/vote/components/DelegateTable/DelegateTable.contracts";
import { translations as voteTranslations } from "@/domains/vote/i18n";
import { data } from "@/tests/fixtures/coins/ark/devnet/delegates.json";
import { env, getDefaultProfileId, render, screen } from "@/utils/testing-library";

let wallet: Contracts.IReadWriteWallet;
let delegate: Contracts.IReadOnlyWallet;

const translations = buildTranslations();

const continueButton = () => screen.getByTestId("DelegateTable__continue-button");

describe("DelegateFooter", () => {
	beforeAll(() => {
		const profile = env.profiles().findById(getDefaultProfileId());
		wallet = profile.wallets().values()[0];

		delegate = new ReadOnlyWallet({
			address: data[0].address,
			explorerLink: "",
			governanceIdentifier: "address",
			isDelegate: true,
			isResignedDelegate: false,
			publicKey: data[0].publicKey,
			username: data[0].username,
		});
	});

	it("should render", () => {
		const { container, asFragment } = render(
			<DelegateFooter
				selectedWallet={wallet}
				availableBalance={wallet.balance()}
				selectedVotes={[]}
				selectedUnvotes={[]}
				maxVotes={wallet.network().maximumVotesPerTransaction()}
			/>,
		);

		expect(container).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should show available balance if network requires vote amount", () => {
		const { rerender } = render(
			<DelegateFooter
				selectedWallet={wallet}
				availableBalance={wallet.balance()}
				selectedVotes={[]}
				selectedUnvotes={[]}
				maxVotes={wallet.network().maximumVotesPerTransaction()}
			/>,
		);

		expect(screen.queryByTestId("DelegateTable__available-balance")).not.toBeInTheDocument();

		const votesAmountMinimumMock = jest.spyOn(wallet.network(), "votesAmountMinimum").mockReturnValue(10);

		rerender(
			<DelegateFooter
				selectedWallet={wallet}
				availableBalance={wallet.balance()}
				selectedVotes={[]}
				selectedUnvotes={[]}
				maxVotes={wallet.network().maximumVotesPerTransaction()}
			/>,
		);

		expect(screen.getByTestId("DelegateTable__available-balance")).toBeInTheDocument();

		votesAmountMinimumMock.mockRestore();
	});

	it("should calculate remaining balance show it", () => {
		const votesAmountMinimumMock = jest.spyOn(wallet.network(), "votesAmountMinimum").mockReturnValue(10);

		render(
			<DelegateFooter
				selectedWallet={wallet}
				availableBalance={wallet.balance() / 2}
				selectedVotes={[]}
				selectedUnvotes={[]}
				maxVotes={wallet.network().maximumVotesPerTransaction()}
			/>,
		);

		expect(screen.getByTestId("DelegateTable__available-balance")).toBeInTheDocument();

		expect(
			screen.getByText(
				translations.VOTE.DELEGATE_TABLE.VOTE_AMOUNT.AVAILABLE_TO_VOTE.replace("{{percent}}", "50"),
			),
		).toBeInTheDocument();

		expect(screen.getByText(`16.87544901 ${wallet.network().ticker()}`)).toBeInTheDocument();

		votesAmountMinimumMock.mockRestore();
	});

	it("should disable continue button with tooltip if user doesn't select a delegate", () => {
		const selectedDelegate: VoteDelegateProperties[] = [
			{
				amount: 0,
				delegateAddress: delegate.address(),
			},
		];

		const { rerender, baseElement } = render(
			<DelegateFooter
				selectedWallet={wallet}
				availableBalance={wallet.balance()}
				selectedVotes={[]}
				selectedUnvotes={[]}
				maxVotes={wallet.network().maximumVotesPerTransaction()}
				onContinue={jest.fn()}
			/>,
		);

		expect(continueButton()).toBeDisabled();

		userEvent.hover(screen.getByTestId("DelegateTable__continue--wrapper"));

		expect(baseElement).toHaveTextContent(voteTranslations.DELEGATE_TABLE.TOOLTIP.SELECTED_DELEGATE);

		rerender(
			<DelegateFooter
				selectedWallet={wallet}
				availableBalance={wallet.balance()}
				selectedVotes={selectedDelegate}
				selectedUnvotes={[]}
				maxVotes={wallet.network().maximumVotesPerTransaction()}
			/>,
		);

		expect(continueButton()).not.toBeDisabled();

		rerender(
			<DelegateFooter
				selectedWallet={wallet}
				availableBalance={wallet.balance()}
				selectedVotes={[]}
				selectedUnvotes={selectedDelegate}
				maxVotes={wallet.network().maximumVotesPerTransaction()}
			/>,
		);

		expect(continueButton()).not.toBeDisabled();

		userEvent.hover(screen.getByTestId("DelegateTable__continue--wrapper"));

		expect(baseElement).not.toHaveTextContent(voteTranslations.DELEGATE_TABLE.TOOLTIP.SELECTED_DELEGATE);
	});

	it("should disable continue button with tooltip if there is at least 1 empty amount field when network requires vote amount", () => {
		const votesAmountMinimumMock = jest.spyOn(wallet.network(), "votesAmountMinimum").mockReturnValue(10);

		const selectedDelegate: VoteDelegateProperties[] = [
			{
				amount: 0,
				delegateAddress: delegate.address(),
			},
		];

		const { rerender, baseElement } = render(
			<DelegateFooter
				selectedWallet={wallet}
				availableBalance={wallet.balance()}
				selectedVotes={selectedDelegate}
				selectedUnvotes={[]}
				maxVotes={wallet.network().maximumVotesPerTransaction()}
				onContinue={jest.fn()}
			/>,
		);

		expect(continueButton()).toBeDisabled();

		userEvent.hover(screen.getByTestId("DelegateTable__continue--wrapper"));

		expect(baseElement).toHaveTextContent(voteTranslations.DELEGATE_TABLE.TOOLTIP.INVALID_AMOUNT);

		rerender(
			<DelegateFooter
				selectedWallet={wallet}
				availableBalance={wallet.balance()}
				selectedVotes={[
					{
						amount: 10,
						delegateAddress: delegate.address(),
					},
				]}
				selectedUnvotes={[]}
				maxVotes={wallet.network().maximumVotesPerTransaction()}
			/>,
		);

		expect(continueButton()).not.toBeDisabled();

		votesAmountMinimumMock.mockRestore();
	});
});
