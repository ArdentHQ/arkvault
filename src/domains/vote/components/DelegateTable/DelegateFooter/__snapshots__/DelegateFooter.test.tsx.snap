// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DelegateFooter > should render 1`] = `
<DocumentFragment>
  <div
    class="fixed inset-x-0 bottom-0 mb-14 h-auto w-screen bg-theme-background py-3 shadow-footer-smooth dark:bg-black dark:shadow-footer-smooth-dark sm:mb-0 md:py-4 lg:py-8"
    data-testid="DelegateTable__footer"
  >
    <div
      class="mx-auto px-8 lg:container md:px-10"
    >
      <div
        class="flex flex-col space-y-2 font-semibold sm:h-11 sm:flex-row sm:space-x-3 sm:space-y-0"
      >
        <div
          class="flex flex-grow divide-theme-secondary-300 overflow-x-auto dark:divide-theme-secondary-800 sm:mr-auto sm:divide-x"
        >
          <div
            class="flex flex-grow overflow-x-auto"
          >
            <div
              class="mr-3 hidden lg:block"
            >
              <div
                class="shrink-0 css-tthe5h"
                data-testid="Avatar"
              >
                <div
                  class="inline-flex h-full w-full items-center justify-center overflow-hidden rounded-full align-middle"
                >
                  <img
                    alt="D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
                    src="data:image/svg+xml;utf8,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" class="picasso" width="100" height="100" viewBox="0 0 100 100"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill="rgb(0, 150, 136)" width="100" height="100"/><circle r="50" cx="80" cy="50" fill="rgb(76, 175, 80)"/><circle r="35" cx="100" cy="40" fill="rgb(156, 39, 176)"/><circle r="40" cx="90" cy="70" fill="rgb(3, 169, 244)"/></svg>"
                    title="D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
                  />
                </div>
              </div>
            </div>
            <div
              class="flex h-full flex-grow flex-col justify-between overflow-x-auto"
            >
              <div
                class="hidden sm:block css-bnrgkd"
              >
                Address
              </div>
              <div
                class="flex h-5 items-center overflow-hidden lg:h-auto lg:space-y-2"
              >
                <div
                  class="flex pr-3 lg:hidden"
                >
                  <div
                    class="shrink-0 css-jxpwdn"
                    data-testid="Avatar"
                  >
                    <div
                      class="inline-flex h-full w-full items-center justify-center overflow-hidden rounded-full align-middle"
                    >
                      <img
                        alt="D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
                        src="data:image/svg+xml;utf8,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" class="picasso" width="100" height="100" viewBox="0 0 100 100"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill="rgb(0, 150, 136)" width="100" height="100"/><circle r="50" cx="80" cy="50" fill="rgb(76, 175, 80)"/><circle r="35" cx="100" cy="40" fill="rgb(156, 39, 176)"/><circle r="40" cx="90" cy="70" fill="rgb(3, 169, 244)"/></svg>"
                        title="D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="lg:text flex w-full overflow-x-auto xl:-mt-px"
                >
                  <div
                    class="flex overflow-hidden whitespace-nowrap items-center space-x-2 w-full"
                  >
                    <span
                      class="font-semibold text-base lg:text-lg !leading-tighter text-theme-text"
                      data-testid="Address__alias"
                    >
                      <span
                        class="no-ligatures css-u3jrna"
                        data-testid="TruncateEnd"
                      >
                        ARK Wallet 1
                      </span>
                    </span>
                    <span
                      class="no-ligatures min-w-0 overflow-hidden lg:text-lg !leading-tighter text-theme-secondary-500 dark:text-theme-secondary-700 font-semibold text-base"
                      data-testid="Address__address"
                    >
                      D8rr7B1d6TL6pf14LgMz4sKp1VBMs6YUYD
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="flex h-full border-t border-theme-secondary-300 pt-2 dark:border-theme-secondary-800 sm:border-l sm:border-t-0 sm:pt-0 md:border-l-0"
        >
          <div
            class="-ml-6 flex divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800 sm:ml-0"
          >
            <div
              class="flex space-x-3 px-6"
            >
              <div
                class="flex flex-col justify-between sm:text-right"
              >
                <div
                  class="css-bnrgkd"
                >
                  Votes
                </div>
                <div
                  class="css-7kkgfg"
                  data-testid="DelegateTable__footer--vote"
                  disabled=""
                >
                  <div
                    class="flex items-center justify-start md:justify-end"
                  >
                    <div>
                      0
                    </div>
                    <div
                      class="hidden pl-2 md:block lg:hidden"
                    >
                      <div
                        class="shrink-0 css-1i4b0eq"
                        height="20"
                        width="20"
                      >
                        <svg
                          id="vote"
                          viewBox="0 0 20 20"
                          x="0"
                          xml:space="preserve"
                          xmlns="http://www.w3.org/2000/svg"
                          y="0"
                        >
                          <g
                            fill="none"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                          >
                            <path
                              d="M7 9.404l2.1 3.182c.2.298.4.398.7.398s.6-.1.8-.299l6.7-9.05"
                            />
                            <path
                              d="M19 12.337v4.077C19 17.807 17.9 19 16.5 19H3.6C2.2 18.9 1 17.906 1 16.414V3.486C1 2.094 2.1 1 3.5 1h6.8"
                            />
                          </g>
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="hidden lg:block"
              >
                <div
                  class="ring-theme-background duration-100 css-188kr7j"
                  disabled=""
                >
                  <div
                    class="css-1i4b0eq"
                    height="20"
                    width="20"
                  >
                    <svg
                      id="vote"
                      viewBox="0 0 20 20"
                      x="0"
                      xml:space="preserve"
                      xmlns="http://www.w3.org/2000/svg"
                      y="0"
                    >
                      <g
                        fill="none"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      >
                        <path
                          d="M7 9.404l2.1 3.182c.2.298.4.398.7.398s.6-.1.8-.299l6.7-9.05"
                        />
                        <path
                          d="M19 12.337v4.077C19 17.807 17.9 19 16.5 19H3.6C2.2 18.9 1 17.906 1 16.414V3.486C1 2.094 2.1 1 3.5 1h6.8"
                        />
                      </g>
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex space-x-3 px-6"
            >
              <div
                class="flex flex-col justify-between sm:text-right"
              >
                <div
                  class="css-bnrgkd"
                >
                  Unvotes
                </div>
                <div
                  class="css-7kkgfg"
                  data-testid="DelegateTable__footer--unvote"
                  disabled=""
                >
                  <div
                    class="flex items-center justify-start md:justify-end"
                  >
                    <div>
                      0
                    </div>
                    <div
                      class="hidden pl-2 md:block lg:hidden"
                    >
                      <div
                        class="shrink-0 css-1i4b0eq"
                        height="20"
                        width="20"
                      >
                        <svg
                          id="unvote"
                          viewBox="0 0 20 20"
                          x="0"
                          xml:space="preserve"
                          xmlns="http://www.w3.org/2000/svg"
                          y="0"
                        >
                          <g
                            fill="none"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                          >
                            <path
                              d="M5.35 14.6l9.2-9.2M14.65 14.6l-9.2-9.2M19 12.337v4.077C19 17.807 17.9 19 16.5 19H3.6C2.2 18.9 1 17.906 1 16.414V3.486C1 2.094 2.1 1 3.5 1h6.8"
                            />
                          </g>
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="hidden lg:block"
              >
                <div
                  class="ring-theme-background duration-100 css-188kr7j"
                  disabled=""
                >
                  <div
                    class="css-1i4b0eq"
                    height="20"
                    width="20"
                  >
                    <svg
                      id="unvote"
                      viewBox="0 0 20 20"
                      x="0"
                      xml:space="preserve"
                      xmlns="http://www.w3.org/2000/svg"
                      y="0"
                    >
                      <g
                        fill="none"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      >
                        <path
                          d="M5.35 14.6l9.2-9.2M14.65 14.6l-9.2-9.2M19 12.337v4.077C19 17.807 17.9 19 16.5 19H3.6C2.2 18.9 1 17.906 1 16.414V3.486C1 2.094 2.1 1 3.5 1h6.8"
                        />
                      </g>
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="flex space-x-3 px-6 hidden md:flex"
            >
              <div
                class="flex flex-col justify-between sm:text-right"
              >
                <div
                  class="css-bnrgkd"
                >
                  Total
                </div>
                <div
                  class="css-7rfz36"
                  data-testid="DelegateTable__footer--votecombination"
                >
                  <div
                    class="flex items-center justify-start md:justify-end"
                  >
                    <div>
                      0/1
                    </div>
                    <div
                      class="hidden pl-2 md:block lg:hidden"
                    >
                      <div
                        class="shrink-0 css-1i4b0eq"
                        height="20"
                        width="20"
                      >
                        <svg
                          id="vote-combination"
                          viewBox="0 0 20 20"
                          x="0"
                          xml:space="preserve"
                          xmlns="http://www.w3.org/2000/svg"
                          y="0"
                        >
                          <g
                            fill="none"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                          >
                            <path
                              d="M5.65 14.24l8.7-8.7M12.15 14.86l2.6-2.7M14.85 14.86l-2.6-2.7M5.15 7.14l.7 1.1c.1.2.2.2.4.1h0l.1-.1 2.4-3.1"
                            />
                            <path
                              d="M19 12.337v4.077C19 17.807 17.9 19 16.5 19H3.6C2.2 18.9 1 17.906 1 16.414V3.486C1 2.094 2.1 1 3.5 1h6.8"
                            />
                          </g>
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="hidden lg:block"
              >
                <div
                  class="ring-theme-background duration-100 css-1ngxrs8"
                >
                  <div
                    class="css-1i4b0eq"
                    height="20"
                    width="20"
                  >
                    <svg
                      id="vote-combination"
                      viewBox="0 0 20 20"
                      x="0"
                      xml:space="preserve"
                      xmlns="http://www.w3.org/2000/svg"
                      y="0"
                    >
                      <g
                        fill="none"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      >
                        <path
                          d="M5.65 14.24l8.7-8.7M12.15 14.86l2.6-2.7M14.85 14.86l-2.6-2.7M5.15 7.14l.7 1.1c.1.2.2.2.4.1h0l.1-.1 2.4-3.1"
                        />
                        <path
                          d="M19 12.337v4.077C19 17.807 17.9 19 16.5 19H3.6C2.2 18.9 1 17.906 1 16.414V3.486C1 2.094 2.1 1 3.5 1h6.8"
                        />
                      </g>
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <span
            class="ml-auto"
            data-testid="DelegateTable__continue--wrapper"
          >
            <button
              class="css-1yqe28v"
              data-testid="DelegateTable__continue-button"
              disabled=""
              type="button"
            >
              <div
                class="flex items-center space-x-2"
              >
                Continue
              </div>
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
